src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -107,7 +129,48 @@ def get_install_requirement(self):
         return self._ireq
 
 
-class ExtrasCandidate(LinkCandidate):
+class InstalledCandidate(Candidate):
+    def __init__(
+        self,
+        dist,  # type: Distribution
+        parent,  # type: InstallRequirement
+        factory,  # type: Factory
+    ):
+        # type: (...) -> None
+        self.dist = dist
+        self._ireq = make_install_req_from_dist(dist, parent)
+        self._factory = factory
+
+        # This is just logging some messages, so we can do it eagerly.
+        # The returned dist would be exactly the same as self.dist because we
+        # set satisfied_by in make_install_req_from_dist.
+        # TODO: Supply reason based on force_reinstall and upgrade_strategy.
+        skip_reason = "already satisfied"
+        factory.preparer.prepare_installed_requirement(self._ireq, skip_reason)
+
+    @property
+    def name(self):
+        # type: () -> str
+        return canonicalize_name(self.dist.project_name)
```

Comment:
Probably not necessary, as `dist.project_name` should always be in canonical form. But no real harm in being certain I guess...

Reply:
I know `pkg_resources` uses a different function to canonicalise names (`safe_name` or something), but I’ve never been able to tell if it’s actually the same as `canonicalize_name` or not 😞

Reply:
I know what you mean... 🙁